adapter/statsd/statsd.go

```diff
@@ -0,0 +1,178 @@
+	flushDuration, err := ptypes.Duration(params.FlushDuration)
+	if err != nil {
+		ce = ce.Append("FlushDuration", err)
+	}
+	if flushDuration < time.Duration(0) {
+		ce = ce.Append("FlushDuration", fmt.Errorf("flush duration must be >= 0"))
+	}
+	if params.FlushBytes < 0 {
+		ce = ce.Append("FlushBytes", fmt.Errorf("flush bytes must be >= 0"))
+	}
+	if params.SamplingRate < 0 {
+		ce = ce.Append("SamplingRate", fmt.Errorf("sampling rate must be >= 0"))
+	}
+	if _, err := net.ResolveUDPAddr("udp", params.Address); err != nil {
+		ce = ce.Append("Address", fmt.Errorf("could not resolve address '%s' with err: %s", params.Address, err))
+	}
+	for metricName, s := range params.MetricNameTemplateStrings {
+		if _, err := template.New(metricName).Parse(s); err != nil {
+			ce = ce.Append("MetricNameTemplateStrings", fmt.Errorf("failed to parse template '%s' for metric '%s' with err: %s", s, metricName, err))
+		}
+	}
+	return
+}
+
+func (*builder) NewMetricsAspect(env adapter.Env, cfg adapter.AspectConfig, metrics []adapter.MetricDefinition) (adapter.MetricsAspect, error) {
+	params := cfg.(*config.Params)
+
+	flushBytes := int(params.FlushBytes)
+	if flushBytes <= 0 {
+		env.Logger().Infof("Got FlushBytes of '%d', defaulting to '%d'", flushBytes, defaultFlushBytes)
+		// the statsd impl we use defaults to 1432 byte UDP packets when flushBytes <= 0; we want to default to 512 so we check ourselves.
+		flushBytes = defaultFlushBytes
+	}
+
+	flushDuration, _ := ptypes.Duration(params.FlushDuration)
+	client, _ := statsd.NewBufferedClient(params.Address, params.Prefix, flushDuration, flushBytes)
+
+	templates := make(map[string]*template.Template)
+	for metricName, s := range params.MetricNameTemplateStrings {
+		def, found := findMetric(metrics, metricName)
+		if !found {
+			env.Logger().Infof("template registered for nonexistent metric '%s'", metricName)
+			continue // we don't have a metric that corresponds to this template, skip processing it
+		}
+
+		t, _ := template.New(metricName).Parse(s)
+		if err := t.Execute(ioutil.Discard, def.Labels); err != nil {
+			env.Logger().Warningf(
+				"skipping custom statsd metric name for metric '%s', could not satisfy template '%s' with labels '%v' with err: %s",
+				metricName, s, def.Labels, err)
+			continue
+		}
+		templates[metricName] = t
+	}
+	return &aspect{params.SamplingRate, client, templates}, nil
+}
+
+func (a *aspect) Record(values []adapter.Value) error {
+	var result *multierror.Error
+	for _, v := range values {
+		if err := a.record(v); err != nil {
+			result = multierror.Append(result, err)
+		}
+	}
+	return result.ErrorOrNil()
+}
+
+func (a *aspect) record(value adapter.Value) error {
+	name := value.Name
+	if t, found := a.templates[value.Name]; found {
+		buf := new(bytes.Buffer)
+		if err := t.Execute(buf, value.Labels); err != nil {
+			panic(fmt.Errorf("failed to create metric name with template '%s' and labels '%v'; got err: %s", t.Name(), value.Labels, err))
```

Comment:
Turns out my understanding of the template code was a bit wrong: `Execute` only throws due to invalid templates (which we check for in ValidateConfig), not due to not being able to satisfy the template with the input.

So for the example of the template `{{.apiMethod}}-{{.responseCode}}`, trying to `Execute` it with a value that has no labels results in the string `<no value>-<no value>`. I'm pretty convinced that this panic is unreachable given validated config. Should we keep it around, just ignore that error, or change it to a log statement and default to the metric's original name?